File: trust-tasks-rs/src/transport.rs
//! [`TransportHandler`] trait — the integration point between the framework's
//! transport-agnostic document model and a concrete transport binding (REST,
//! DIDComm, AMQP, ...).
//!
//! The contract encodes SPEC.md §4.8.1 and §9.2:
//!
//! - In-band `issuer` / `recipient` values are authoritative when present.
//! - Transport-derived identity fills in absent members.
//! - When both are present, they MUST be consistent — a mismatch is a
//!   validation failure (the standard error code is `identity_mismatch`,
//!   §8.3).
//!
//! A handler implementation supplies the *transport-derived* values it
//! observed for an inbound message via [`TransportHandler::derive_parties`];
//! the [`TransportHandler::resolve_parties`] default method applies the
//! precedence rule to produce a single [`ResolvedParties`] value that the
//! consumer applies for every subsequent framework rule.

use serde::Serialize;
use thiserror::Error;

use crate::document::{ErrorResponse, TrustTask};
use crate::error::RejectReason;

/// The result of resolving a document's in-band party identity against the
/// transport context.
///
/// Per SPEC.md §4.8.1, these are the values a consumer applies for every
/// subsequent framework rule that references the issuer or recipient.
#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct ResolvedParties {
    /// The party responsible for the document's content, after precedence.
    pub issuer: Option<String>,
    /// The party expected to act upon the document, after precedence.
    pub recipient: Option<String>,
}

/// What the transport handler observed about an inbound message's parties.
///
/// Returned by [`TransportHandler::derive_parties`]. A field is `None` when
/// the transport does not authenticate that party — for example, a plain
/// HTTPS handler with no client certificate sets `issuer = None`.
#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct TransportContext {
    /// Transport-authenticated sender identity, if any.
    pub issuer: Option<String>,
    /// Transport-authenticated recipient identity, if any.
    pub recipient: Option<String>,
}

/// Errors raised when in-band and transport-derived identity disagree
/// (SPEC.md §7.2 item 6; standard code `identity_mismatch`, §8.3).
#[derive(Debug, Error, Clone, PartialEq, Eq)]
pub enum ConsistencyError {
    /// The in-band `issuer` does not match the transport-authenticated sender.
    #[error("in-band issuer {in_band:?} does not match transport-derived {transport:?}")]
    IssuerMismatch {
        /// Value carried in the document's `issuer` member.
        in_band: String,
        /// Value derived from the transport.
        transport: String,
    },
    /// The in-band `recipient` does not match the transport-authenticated
    /// addressee.
    #[error("in-band recipient {in_band:?} does not match transport-derived {transport:?}")]
    RecipientMismatch {
        /// Value carried in the document's `recipient` member.
        in_band: String,
        /// Value derived from the transport.
        transport: String,
    },
}

/// A transport binding's plug-in for the framework.
///
/// Implementations represent one transport (REST + mTLS, DIDComm, TSP, an
/// in-memory test loopback, ...) and supply the transport-derived information
/// the framework needs to apply SPEC.md §4.8.1 precedence.
///
/// The trait is sync and object-safe; transports that perform I/O do it
/// outside the trait (e.g. before constructing the handler with the values it
/// already extracted from a verified envelope).
pub trait TransportHandler {
    /// A stable identifier for this transport binding, suitable for logs and
    /// audit (SPEC.md §9.1, §9.2 — "A *transport binding* specification
    /// SHOULD identify itself by a stable URI").
    fn binding_uri(&self) -> &str;

    /// Identities the transport authenticated for the inbound message under
    /// consideration.
    ///
    /// Implementations construct the handler with whatever the transport
    /// extracted (peer certificate subject, verified DIDComm sender,
    /// authenticated routing key, ...) and return it here. Returning a
    /// fully-empty [`TransportContext`] is valid for transports that provide
    /// no authenticated identity — the framework then falls back entirely to
    /// the in-band members and any `proof` they carry.
    fn derive_parties(&self) -> TransportContext;

    /// Prepare an outbound document for emission over this transport.
    ///
    /// Per SPEC.md §9.2 item 1, a producer-side handler MAY strip `issuer` /
    /// `recipient` when the transport will provide authenticated identity
    /// for those roles end-to-end. The default implementation leaves the
    /// document untouched, which is the conservative choice (in-band values
    /// remain visible to downstream consumers regardless of transport).
    fn prepare_outbound<P: Serialize>(&self, _doc: &mut TrustTask<P>) {}

    /// Apply SPEC.md §4.8.1 precedence to produce the final
    /// [`ResolvedParties`] for an inbound document.
    ///
    /// Returns [`ConsistencyError`] when an in-band member is present and
    /// disagrees with the transport-derived value for the same party. The
    /// caller is responsible for translating the error into an
    /// `identity_mismatch` error response (SPEC.md §8.3).
    fn resolve_parties<P>(&self, doc: &TrustTask<P>) -> Result<ResolvedParties, ConsistencyError> {
        let ctx = self.derive_parties();

        let issuer = match (doc.issuer.as_deref(), ctx.issuer.as_deref()) {
            (Some(in_band), Some(transport)) if in_band != transport => {
                return Err(ConsistencyError::IssuerMismatch {
                    in_band: in_band.to_string(),
                    transport: transport.to_string(),
                });
            }
            (Some(in_band), _) => Some(in_band.to_string()),
            (None, Some(transport)) => Some(transport.to_string()),
            (None, None) => None,
        };

        let recipient = match (doc.recipient.as_deref(), ctx.recipient.as_deref()) {
            (Some(in_band), Some(transport)) if in_band != transport => {
                return Err(ConsistencyError::RecipientMismatch {
                    in_band: in_band.to_string(),
                    transport: transport.to_string(),
                });
            }
            (Some(in_band), _) => Some(in_band.to_string()),
            (None, Some(transport)) => Some(transport.to_string()),
            (None, None) => None,
        };

        Ok(ResolvedParties { issuer, recipient })
    }

    /// Build a `trust-task-error/0.1` response for `doc` that satisfies the
    /// SPEC.md §8.1 routing rules — most importantly, the rule that under
    /// [`RejectReason::IdentityMismatch`] the error MUST address the
    /// transport-authenticated sender and MUST NOT address the contested
    /// in-band issuer.
    ///
    /// Returns `Some(ErrorResponse)` when a routable recipient exists, and
    /// `None` when the rejection is `identity_mismatch` and the transport
    /// has no authenticated sender for the inbound document — per §8.1, the
    /// consumer SHOULD NOT emit a response in that case.
    fn reject<P>(
        &self,
        doc: &TrustTask<P>,
        id: impl Into<String>,
        reason: RejectReason,
    ) -> Option<ErrorResponse> {
        let recipient = match &reason {
            RejectReason::IdentityMismatch(_) => {
                // §8.1: transport-authenticated sender only. If the
                // transport authenticated nothing, the consumer suppresses
                // the response.
                let from_transport = self.derive_parties().issuer;
                from_transport.as_ref()?;
                from_transport
            }
            _ => doc.issuer.clone(),
        };
        Some(doc.reject_with_recipient(id, reason, recipient))
    }
}

File: trust-tasks-go/trusttasks/transport.go
package trusttasks

// The integration point between the transport-agnostic document model and a
// concrete transport binding (HTTPS + mTLS, DIDComm, TSP, an in-memory test
// loopback, …).
//
// Mirrors transport.rs in trust-tasks-rs and transport.ts in
// @openvtc/trust-tasks. The contract encodes SPEC.md §4.8.1 and §9.2:
//
//   - In-band issuer / recipient are authoritative when present.
//   - Transport-derived identity fills in absent members.
//   - When both are present they MUST agree; a mismatch is a validation failure
//     (identityMismatch, §8.3).

import "fmt"

// ResolvedParties is party identity after §4.8.1 precedence — the values a
// consumer applies for every subsequent framework rule referencing the issuer or
// recipient.
type ResolvedParties struct {
	Issuer    *string
	Recipient *string
}

// TransportContext is what the transport authenticated about an inbound message.
//
// A field is nil when the transport does not authenticate that party — a plain
// HTTPS handler with no client certificate leaves Issuer unset. Returning an
// entirely empty context is valid; the framework then relies on the in-band
// members and any proof they carry.
type TransportContext struct {
	Issuer    *string
	Recipient *string
}

// ConsistencyError is raised when in-band and transport-derived identity
// disagree (§7.2 item 6).
type ConsistencyError struct {
	// Party is "issuer" or "recipient".
	Party     string
	InBand    string
	Transport string
}

func (e *ConsistencyError) Error() string {
	// Local diagnostics only. The wire message is [IdentityMismatchReason]'s,
	// which deliberately names neither value — see the note there.
	return fmt.Sprintf(
		"trusttasks: in-band %s %q does not match transport-derived %q",
		e.Party, e.InBand, e.Transport,
	)
}

// TransportHandler is a transport binding's plug-in for the framework.
type TransportHandler interface {
	// BindingURI is a stable identifier for this binding, for logs and audit
	// (§9.1, §9.2).
	BindingURI() string

	// DeriveParties reports the identities the transport authenticated for the
	// message under consideration.
	DeriveParties() TransportContext
}

// ResolveParties applies §4.8.1 precedence to produce the final
// [ResolvedParties].
//
// Returns a [ConsistencyError] when an in-band member is present and disagrees
// with the transport-derived value for the same party. Callers translate that
// into an identityMismatch error response (§8.3) — or let [ConsumeInbound] do it.
func ResolveParties[P any](
	handler TransportHandler,
	doc *Document[P],
) (ResolvedParties, *ConsistencyError) {
	ctx := handler.DeriveParties()

	for _, pair := range []struct {
		party     string
		inBand    *string
		transport *string
	}{
		{"issuer", doc.Issuer, ctx.Issuer},
		{"recipient", doc.Recipient, ctx.Recipient},
	} {
		if pair.inBand != nil && pair.transport != nil && *pair.inBand != *pair.transport {
			return ResolvedParties{}, &ConsistencyError{
				Party:     pair.party,
				InBand:    *pair.inBand,
				Transport: *pair.transport,
			}
		}
	}

	parties := ResolvedParties{Issuer: doc.Issuer, Recipient: doc.Recipient}
	if parties.Issuer == nil {
		parties.Issuer = ctx.Issuer
	}
	if parties.Recipient == nil {
		parties.Recipient = ctx.Recipient
	}
	return parties, nil
}

// Reject builds the error response for doc, applying the §8.1 routing rules.
//
// Returns nil when the rejection is identityMismatch and the transport
// authenticated no sender. §8.1 is explicit that the consumer SHOULD NOT emit a
// response in that case: the in-band issuer is by definition the contested
// identity, so addressing it would be an oracle, and on any transport that signs
// error responses it would compel the consumer to emit a signed document about a
// party that did not take part in the exchange.
func Reject[P any](
	handler TransportHandler,
	doc *Document[P],
	id string,
	reason RejectReason,
	clock Clock,
) *ErrorResponse {
	var recipient *string
	if reason.Code == CodeIdentityMismatch {
		recipient = handler.DeriveParties().Issuer
		if recipient == nil {
			return nil
		}
	} else {
		recipient = doc.Issuer
	}
	return RejectWithRecipient(doc, id, ToErrorPayload(reason), recipient, clock)
}

// IdentityMismatchReason is the identityMismatch reason for a
// [ConsistencyError].
//
// §8.1 additionally requires the wire message to be sanitized: naming the
// consumer's expected transport-authenticated identity, or echoing the contested
// in-band value, leaks identity information to a possibly hostile sender. The
// standard wire form is the code alone with a non-identifying message — so the
// mismatched values are deliberately not included here. Log them locally
// instead; [ConsistencyError.Error] carries them.
func IdentityMismatchReason(_ *ConsistencyError) RejectReason {
	return RejectReason{
		Code:    CodeIdentityMismatch,
		Message: "identityMismatch: in-band identity does not match transport-derived identity",
	}
}

// UnauthenticatedTransport is a handler for transports that authenticate nothing
// — an unauthenticated HTTP POST, a public queue, paper. Party identity comes
// entirely from the in-band members and whatever proof they carry.
type UnauthenticatedTransport struct {
	// URI overrides the reported binding URI. Empty means the default below.
	URI string
}

// BindingURI implements [TransportHandler].
func (t UnauthenticatedTransport) BindingURI() string {
	if t.URI == "" {
		return "urn:trust-tasks:transport:unauthenticated"
	}
	return t.URI
}

// DeriveParties implements [TransportHandler], authenticating nothing.
func (t UnauthenticatedTransport) DeriveParties() TransportContext {
	return TransportContext{}
}

// StaticTransport is a fixed-identity handler, for tests and for transports
// resolved out-of-band.
type StaticTransport struct {
	Context TransportContext
	// URI overrides the reported binding URI. Empty means the default below.
	URI string
}

// BindingURI implements [TransportHandler].
func (t StaticTransport) BindingURI() string {
	if t.URI == "" {
		return "urn:trust-tasks:transport:static"
	}
	return t.URI
}

// DeriveParties implements [TransportHandler].
func (t StaticTransport) DeriveParties() TransportContext {
	return t.Context
}
